sql/rowexec/dml.go

```diff
@@ -101,9 +101,15 @@ func (b *BaseBuilder) buildInsertInto(ctx *sql.Context, ii *plan.InsertInto, row
 	}
 
 	if ii.Ignore {
+		// If ignore is set, then we are either replacing or inserting, but not updating on conflicts
 		return plan.NewCheckpointingTableEditorIter(insertIter, ed), nil
 	} else {
-		return plan.NewTableEditorIter(insertIter, ed), nil
+		// Otherwise, we are potentially inserting AND updating if there are conflicts
+		eds := []sql.EditOpenerCloser{ed}
+		if updater != nil {
+			eds = append(eds, updater)
+		}
+		return plan.NewTableEditorIter(insertIter, eds...), nil
```

Comment:
AFAICT we never use more than one editor in practice, despite this constructor accepting multiple editors. I guess if it works and the tests pass it's fine?

But it seems like there should be at least one engine test that doesn't work before this change and passes now.

Reply:
`insertIter` takes a replacer, an updater, and an inserter. If the replacer isn't set, it'll use the inserter, and if the inserter fails, it'll handle an `ON DUPLICATE` clause and use the updater. 

This doesn't repro with a regular, user table, but is triggered by the `docWriter` implementation (specifically, the one held in the updater field), because it needs `StatementBegin()` called to initialize its table writer. Without this, there's a panic from trying to use the nil table writer.